PR: https://github.com/odoo/odoo/pull/78696

From: f0b7600314fee4e99de7ff48c91132f2754a9a40
From: Mitul Shah
Date: 2022-06-24 11:49:07

Structural Changes: 2
Total Changes: 37

[IMP] base, crm: set salesperson and team on partner from parent

This commit does below improvements:

1/ If salesperson and sales team are set on an individual partner, we propagate
   those to parent company being created from m2o of the form view.
2/ If the existing parent company is manually linked to an individual partner
   being created, and salesperson and sales team are not set on the individual,
   we set those values from the linked parent (if set on the parent).

Task-2636290

closes odoo/odoo#78696

Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/odoo/addons/base/models/res_partner.py
+++ b/odoo/addons/base/models/res_partner.py
@@ -168,8 +168,12 @@ class Partner(models.Model):
                                "Anywhere else, time values are computed according to the time offset of your web client.")
 
     tz_offset = fields.Char(compute='_compute_tz_offset', string='Timezone offset', invisible=True)
-    user_id = fields.Many2one('res.users', string='Salesperson',
-      help='The internal user in charge of this contact.')
+    user_id = fields.Many2one(
+        'res.users', string='Salesperson',
+        compute='_compute_user_id',
+        precompute=True,  # avoid queries post-create
+        readonly=False, store=True,
+        help='The internal user in charge of this contact.')
     vat = fields.Char(string='Tax ID', index=True, help="The Tax Identification Number. Complete it if the contact is subjected to government taxes. Used in some legal statements.")
     same_vat_partner_id = fields.Many2one('res.partner', string='Partner with same Tax ID', compute='_compute_same_vat_partner_id', store=False)
     same_company_registry_partner_id = fields.Many2one('res.partner', string='Partner with same Company Registry', compute='_compute_same_vat_partner_id', store=False)
@@ -304,6 +308,12 @@ class Partner(models.Model):
         for partner in self:
             partner.tz_offset = datetime.datetime.now(pytz.timezone(partner.tz or 'GMT')).strftime('%z')
 
+    @api.depends('parent_id')
+    def _compute_user_id(self):
+        """ Synchronize sales rep with parent if partner is a person """
+        for partner in self.filtered(lambda partner: not partner.user_id and partner.company_type == 'person' and partner.parent_id.user_id):
+            partner.user_id = partner.parent_id.user_id
+
     @api.depends('user_ids.share', 'user_ids.active')
     def _compute_partner_share(self):
         super_partner = self.env['res.users'].browse(SUPERUSER_ID).partner_id

--- a/odoo/addons/base/tests/test_res_partner.py
+++ b/odoo/addons/base/tests/test_res_partner.py
@@ -157,3 +157,23 @@ class TestPartner(TransactionCase):
         self.assertFalse(self.env.ref('base.public_user').active)
         self.assertFalse(self.env.ref('base.public_partner').active)
         self.assertTrue(self.env.ref('base.public_partner').is_public)
+
+    def test_onchange_parent_sync_user(self):
+        company_1 = self.env['res.company'].create({'name': 'company_1'})
+        test_user = self.env['res.users'].create({
+            'name': 'This user',
+            'login': 'thisu',
+            'email': 'this.user@example.com',
+            'company_id': company_1.id,
+            'company_ids': [company_1.id],
+        })
+        test_parent_partner = self.env['res.partner'].create({
+            'company_type': 'company',
+            'name': 'Micheline',
+            'user_id': test_user.id,
+        })
+        with Form(self.env['res.partner']) as partner_form:
+            partner_form.parent_id = test_parent_partner
+            partner_form.company_type = 'person'
+            partner_form.name = 'Philip'
+            self.assertEqual(partner_form.user_id, test_parent_partner.user_id)

--- a/odoo/addons/base/views/res_partner_views.xml
+++ b/odoo/addons/base/views/res_partner_views.xml
@@ -84,6 +84,7 @@
                     <field name="is_company" invisible="1"/>
                     <field name="type" invisible="1"/>
                     <field name="avatar_128" invisible="1"/>
+                    <field name="user_id" invisible="1"/>
                     <field name="image_1920" widget='image' class="oe_avatar" options='{"preview_image": "avatar_128"}'/>
                     <div class="oe_title">
                         <field name="company_type" options="{'horizontal': true}" widget="radio" groups="base.group_no_one"/>
@@ -94,7 +95,7 @@
                         <field name="parent_id"
                             widget="res_partner_many2one"
                             placeholder="Company Name..."
-                            domain="[('is_company', '=', True)]" context="{'default_is_company': True, 'show_vat': True}"
+                            domain="[('is_company', '=', True)]" context="{'default_is_company': True, 'show_vat': True, 'default_user_id': user_id}"
                             attrs="{'invisible': [('is_company','=', True)]}"/>
                     </div>
                     <group>
